tla/consistency/SingleNode.tla

```diff
@@ -0,0 +1,122 @@
+            \* The ledger entry index is the sequence number
+            /\ ledgerViews[view][seqnum].view \in Views
+            /\ ledgerViews[view][seqnum].tx \in Txs
+
+\* In this abstract version of CCF's consensus layer, each ledger is append-only
+LedgersMonoProp ==
+    [][\A view \in DOMAIN ledgerViews: IsPrefix(ledgerViews[view], ledgerViews[view]')]_ledgerViews
+
+vars == << history, ledgerViews >>
+
+TypeOK ==
+    /\ HistoryTypeOK
+    /\ LedgerTypeOK
+
+Init ==
+    /\ history = <<>>
+    /\ ledgerViews = [ x \in {1} |-> <<>>]
+
+IndexOfLastRequested ==
+    SelectLastInSeq(history, LAMBDA e : e.type \in {RwTxRequest, RoTxRequest})
+
+NextRequestId ==
+    IF IndexOfLastRequested = 0 THEN 0 ELSE history[IndexOfLastRequested].tx+1
+
+\* Submit new read-write transaction
+\* This could be extended to add a notion of session and then check for session consistency
+RwTxRequestAction ==
+    /\ Len(history) < HistoryLimit
+    /\ history' = Append(
+        history, 
+        [type |-> RwTxRequest, tx |-> NextRequestId]
+        )
+    /\ UNCHANGED ledgerViews
+
+\* Execute a read-write transaction
+RwTxExecuteAction ==
+    /\ \E i \in DOMAIN history :
+        /\ history[i].type = RwTxRequest
+        \* Check transaction has not already been added a ledger
+        /\ \A view \in DOMAIN ledgerViews: 
+            {seqnum \in DOMAIN ledgerViews[view]: 
+                history[i].tx = ledgerViews[view][seqnum].tx} = {}
+        \* Note that a transaction can be added to any ledger, simulating the fact
+        \* that it can be picked up by the current leader or any former leader
+        /\ \E view \in DOMAIN ledgerViews:
+                ledgerViews' = [ledgerViews EXCEPT ![view] = 
+                    Append(@,[view |-> view, tx |-> history[i].tx])]
+        /\ UNCHANGED history
+
+\* Response to a read-write transaction
+RwTxResponseAction ==
+    /\ Len(history) < HistoryLimit
+    /\ \E i \in DOMAIN history :
+        \* Check request has been received and executed but not yet responded to
+        /\ history[i].type = RwTxRequest
+        /\ {j \in DOMAIN history: 
+            /\ j > i 
+            /\ history[j].type = RwTxResponse
+            /\ history[j].tx = history[i].tx} = {}
+        /\ \E view \in DOMAIN ledgerViews:
+            /\ \E seqnum \in DOMAIN ledgerViews[view]: 
+                /\ history[i].tx = ledgerViews[view][seqnum].tx
+                /\ history' = Append(
+                    history,[
+                        type |-> RwTxResponse, 
+                        tx |-> history[i].tx, 
+                        observed |-> [x \in 1..seqnum |-> ledgerViews[view][x].tx],
+                        tx_id |-> <<ledgerViews[view][seqnum].view, seqnum>>] )
+    /\ UNCHANGED ledgerViews
+
+\* Sending a committed status message
+\* Note that a request could only be committed if it's in the highest view's ledger
+StatusCommittedResponseAction ==
+    /\ Len(history) < HistoryLimit
```

Comment:
These length constraints feel like a property of a config/run, rather than the spec itself. Is it possible to remove `HistoryLimit` from the spec, and cap `Len(history)` from the `.cfg` or CLI?

Reply:
I've moved these to separate MC files